### DISCUSSION TITLE
Suzuka DA Migrations Format

### DISCUSSION THREAD

**Discussion by l-monninger** [movementlabsxyz/movement#470] [Ideas]
# Summary
In order to allow for safe upgrades, data model changes, and DA migrations, it has been proposed that we provide a data format which describes the canonical state of the DA over time. This would be hardfork migration approach to upgrades. In a simple flexible form, this would be a list of tuples of (a) block height range  tuples indexing (b) the following fields:

- DA namespace/identifier
- DA transaction format version 
- DA block format version
- Transaction ordering rule version
- Full node version (generally)
- Movement Aptos version (framework and node)
- Aptos state model version (more specifically)

**Comment by mzabaluev** [comment]
> block height range

I think the scalar height key by which these states become normative is more maintainable. We will not have gaps in coverage, so the next entry's height implicitly defines the last height for which the preceding entry applies.

> the following fields:

Should all of these fields be enforced? I'd like to avoid overengineering here.

> * DA transaction format version
> * DA block format version
> * Transaction ordering rule version

Makes sense.

> * Full node version (generally)
> * Movement Aptos version (framework and node)

For these, we would need something like semver minimal version spec. Also introspection into the version information.
What is "Movement Aptos node version", BTW? Should it reflect upstream feature parity?

> Aptos state model version (more specifically)

Where is this defined?

**Reply by l-monninger** [reply]
> I think the scalar height key by which these states become normative is more maintainable. We will not have gaps in coverage, so the next entry's height implicitly defines the last height for which the preceding entry applies.

In theory, with Celestia and similar DAs, there is an edge case where we have some parts of the block height one DA and some parts on another. Right now, there is no reason this would actually emerge, but perhaps under some of the transaction fairness block-merging proposals in [MD-7](https://github.com/movementlabsxyz/MIP/pull/7) it could.

We could assume that we would always govern the chain such this would never be the case, but perhaps something like the below is more helpful...

**Reply by l-monninger** [reply]
> Should all of these fields be enforced? I'd like to avoid overengineering here.

>> DA transaction format version
>> DA block format version
>> Transaction ordering rule version

A couple perspectives here...


## Backwards Compatibility, Hard Forks, and Mapping to Implementation
So, in theory, the more expressive we are about the changes that are consequential, the easier it can be to define cases where we have backwards compatibility and allow node operators so leeway in upgrading. However, we have said the we are allowed to--and in fact have the end objective of--doing everything as a **hard fork**. If you take that to the extreme, really all you need is to track the total full-node version over time.

However, I think this format has to be at least partially concerned with defining these shifts in such a way that maps nicely onto implementation the underlying software. 

Furthermore there is a chance for bit of a behavior-value distinction here that we need to contend with. The behavior w.r.t. data formats, transactions batching, etc. could be the same at the same block heights for two networks. But, perhaps the actual DA endpoint, e.g., Celestia network, is different. 

## Versioning the Format Itself
I think we can be more reductive if we allow for versioning the format and elements thereof itself. 

If we have...
```rust
pub struct DaSpecificationV1 {
  full_node_version : HashValue,
  celestia_namespace : Namespace
}

pub struct DaSpecificationV2 {
  full_node_version : HashValue,
  celestia_namespace : Namespace,
  block_compression_format : BlockCompressionFormat
}

pub enum DaSpecification {
   V1((u64, DaSpecificationV1)),
   V2((Range(u64), DaSpecificationV2))
}
```
...that should an approach we can adopt pretty well.

## Mapping to Critical Areas of Code
Continuing from the above, I think it could be particularly maintainable to, at various points, identify areas potential affecting compatibility and extract them into a structure with a serializable representation. We then serialize and deserialize that struct from the config. 

For example, we currently have serialization of blocks for the DA of a particular form that is only represented within the implementation on the Block object. We could instead extract an enum like so:

```rust
pub struct BlockSerializationV1; // might not need fields

impl BlockSerializationV1 {
   // ...
}

pub enum BlockSerialization {
    V1(BlockSerializationV1)
}
```

This can also be defined on the structures being serialized themselves, i.e., with those being versioned. This might be a little rigid, though at the same time more naturally accounts for changes in the underlying format.

These considerations might rear their heads sooner than expected. For example, we have to validate that the current hash being used in memseq with the `BTreeSet` for ordering  s a CSPRNG to insure cryptographic transaction fairness.

**Reply by l-monninger** [reply]
> Where is this defined?

I don't think they actually a very cohesive approach here, so maybe just the Aptos version itself makes sense. Per the above, maybe we should only worry about versions of things in our code and then lump the Aptos version in with our introspective full-node version.

**Reply by mzabaluev** [reply]
So at the V1 definition, we can start with the namespace identifier, the transaction and block format versions, the ordering rule version, and the node minimum version to cover for potential behavioral differences. If anything else emerges that we can differentiate on, this can be added in later format revisions.

I also think that the versioned enum approach is too rigid and creates a lot of unnecessary boilerplate; from experience with dealing with format changes in CometBFT, I would rather take the backward compatibility approach taken in protobuf and many JSON schemas, where fields are not renamed until necessary (though they can be deprecated), and added fields receive default values when deserializing an earlier version of the format.

**Reply by mzabaluev** [reply]
> In theory, with Celestia and similar DAs, there is an edge case where we have some parts of the block height one DA and some parts on another. Right now, there is no reason this would actually emerge, but perhaps under some of the transaction fairness block-merging proposals in https://github.com/movementlabsxyz/MIP/pull/7 it could.

I think it would be simpler to demand that the nodes only produced blobs with DA settings corresponding to the current block height, as defined by the same migration table (which in the future should be part of the governed chain parameters).

**Reply by l-monninger** [reply]
> I also think that the versioned enum approach is too rigid 

Can you elaborate? I

 don't mind the backwards compatibility approach, it mainly just stands out less in code and can create some awkward fields later. 

I suppose the other reason the versioned enum sounds nice to me is that it's easier to reference in governance discussions. That is, you can say something like, "upon adoption of this proposal, the canonical chain uses DA Specification V1."

**Reply by l-monninger** [reply]
> I think it would be simpler to demand that the nodes only produced blobs with DA settings corresponding to the current block height, as defined by the same migration table (which in the future should be part of the governed chain parameters).

Okay, we probably should fly the flag of this approach somewhere s.t. we don't do something stupid down the road.

**Reply by mzabaluev** [reply]
> Can you elaborate?

My thoughts on this have evolved a bit, and I think it's OK to maintain versioned rigid formats for transactions and blobs and use the BCS encoding.

For the migrations config itself though, I would prefer not to get carried away with versioning, besides making the current version a top-level parameter to enable breaking changes in the future.

**Reply by l-monninger** [reply]
> BCS encoding

What is the emphasis on the encoding here?

**Reply by l-monninger** [reply]
> I would prefer not to get carried away with versioning, besides making the current version a top-level parameter to enable breaking changes in the future.

So does this mean using versioned enums, but not worrying too much about the internal structure? I think that's kind of the point of the versioned approach in general.

**Reply by mzabaluev** [reply]
> > I would prefer not to get carried away with versioning, besides making the current version a top-level parameter to enable breaking changes in the future.
> 
> So does this mean using versioned enums, but not worrying too much about the internal structure? I think that's kind of the point of the versioned approach in general.

Essentially a hybrid: put in a `"version": 1` field to be future-proof, but evolve the structure backward-compatibly wrt. adding new fields without bumping the version as long as we can get away with it. This way, we don't need versioned enums until we actually have to introduce version 2.

**Reply by l-monninger** [reply]
> Essentially a hybrid: put in a "version": 1 field to be future-proof, but evolve the structure backward-compatibly wrt. adding new fields without bumping the version as long as we can get away with it. This way, we don't need versioned enums until we actually have to introduce version 2.

Interesting. So, you are trying to avoid enums in code not their implication in the format? Because for most formats essentially how Serde will serialize an enum is some variant of this. For example, with JSON it names the field containing the data according to the version.

**Reply by mzabaluev** [reply]
While we are on version 1, it's possible to serialize and deserialize the schema with plain structs.
Even later, you can parse into a `serde_json::Value`, check the top-level `version` field, hen convert the items into struct instances of a version-appropriate type with `serde_json::from_value`.

**Comment by mzabaluev** [comment]
Something else that we should think about is the location of the migration table and requirements arising out of this.

* Well-known configuration file for the initial governance phase: preferably human-readable, likely JSON.
* On-chain parameter: should it be a binary format that allows extensibility (which excludes BCS)? Protobuf? Do we have a general approach for chain in Movement?

**Reply by l-monninger** [reply]
In the beginning, this will mainly be just for node operators. I think the initial governance suggestion that this is just a human-readable JSON stored somewhere is perfect. Perhaps it can be MIP that houses this in some manner.

**Reply by l-monninger** [reply]
@mzabaluev do you maybe want to open an MD for this?: https://github.com/movementlabsxyz/MIP

This will be part of a governance lineage, so it would be good to have there.

**Comment by mzabaluev** [comment]
Drafting the data types and rules of their application on the basis of the discussion.

## DA Settings

The format of a well-known file defining migrations should be JSON serialization of a map where keys indicate the block height at which the configuration in the value starts applying. The values are objects. The top-level configuration envelope specifies the current version of the format to permit breaking changes in the future.

```json
{
  "version": 1,
  "migrations": {
    "123456": {
      "namespace": "deadbeef01234567890",
      "transaction_format": 1,
      "block_format": 2,
      "ordering_rule": 3,
      "node_version": "1.2"
    }
  }
}
```

the Rust definition of the value struct can look like this:

```rust
#[derive(Serialize, Deserialize)]
pub struct DASettings {
    namespace: Namespace,
    transaction_format: TransactionFormatVersion,
    block_format: BlockFormatVersion,
    ordering_rule: OrderingRuleVersion,
    node_version: NodeVersionReq,
}
```

## Principles of introducing migrations

Each new entry in the migration table SHALL use a unique Celestia namespace identifier to prevent confusion and detect accidental misconfiguration early.

The configuration in all nodes in the network should be updated well in advance of the block height at which a new change is introduced. The expected process in governance stage 0 is a coordinated upgrade.

The `node_version` field gives the semver minimum version of the Suzuka full node software that is required to use these settings, in the "major.minor" format. A node that does not satisfy the version requirement for any of the encountered entries MUST reject the configuration and report an error.

## Applying the migrations

The producers and consumers of DA blobs must implement all versions encountered in the configuration file and apply the settings accordingly to the block height. There is not to be mixing of different DA formats between blobs submitted by correct nodes for the same block height.

## Evolving the migration format

JSON objects are self-describing and allow adding new fields in a backward-compatible way, should we need them later.

For the on-chain parameter to be introduced with later governance stages, we can consider retaining the JSON format, or if a binary encoding is required, use an extensible encoding such as Protobuf.

**Reply by l-monninger** [reply]
Yeah, so apart from the questions about the versioning I still have from above, I think this is good. 

I would basically copy over to an MIP, polish a bit, then look to assign issues.

https://github.com/movementlabsxyz/MIP

**Reply by mzabaluev** [reply]
MIP-13: https://github.com/movementlabsxyz/MIP/pull/13
